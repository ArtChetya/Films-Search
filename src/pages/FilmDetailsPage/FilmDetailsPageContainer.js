import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { httpService } from '../../services/httpService';
import { FilmDetailsPage } from '.';

export const FilmDetailsPageContainer = ({ match }) => {
    const defaultParams = {
        search: undefined,
        searchBy: 'genres',
        sortBy: 'release_date',
        sortOrder: 'desc',
        limit: 50
    };
    const [params] = useState(defaultParams);

    const [filmDetails, setFilmDetails] = useState({});
    const [isFilmDetailsLoading, setIsFilmDetailsLoading] = useState(false);

    useEffect(() => {
        const fetchFilm = async () => {
            setIsFilmDetailsLoading(true);
            const response = await httpService({
                method: 'GET',
                url: `movies/${match.params.id}`
            });

            const film = response.data;
            setFilmDetails(film);
            setIsFilmDetailsLoading(false);
        };

        fetchFilm();
    }, [match.params.id]);

    const [filmsByGenres, setFilmsByGenres] = useState([]);
    const [isFilmsLoading, setIsFilmsLoading] = useState(false);
    const isInitRef = useRef(true);

    useEffect(() => {
        const fetchFilms = async () => {
            if (!isInitRef.current) {
                setIsFilmsLoading(true);
                const response = await httpService({
                    params: {
                        ...params,
                        search: filmDetails.genres[0]
                    },
                    method: 'GET',
                    url: 'movies'
                });

                const { data } = response.data;
                setFilmsByGenres(data);
                setIsFilmsLoading(false);
            } else {
                isInitRef.current = false;
            }
        };

        fetchFilms();
    }, [params, filmDetails]);

    return (
        <FilmDetailsPage
            filmDetails={filmDetails}
            filmsByGenres={filmsByGenres}
            isFilmDetailsLoading={isFilmDetailsLoading}
            isFilmsLoading={isFilmsLoading}
        />
    );
};

FilmDetailsPageContainer.propTypes = {
    match: PropTypes.object.isRequired
};
