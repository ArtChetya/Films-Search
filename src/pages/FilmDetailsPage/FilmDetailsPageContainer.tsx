import { Loader } from 'components';
import { fetchFilmDetailsInfo } from 'features/filmDetails';
import { params as paramsAction } from 'features/searchParams';
import { serverSideRenderedFlag } from 'features/serverSideRendered';
import React, { FunctionComponent, useEffect } from 'react';
import { RouteComponentProps } from 'react-router';
import { FilmDetailsPage } from '.';
import {
    IFilmDetailsPageDispatchProps,
    IFilmDetailsPageStateProps
} from './FilmDetailsPageConnector';

interface IParams {
    id: string;
}

interface IFilmDetailsPageContainerProps
    extends RouteComponentProps<IParams>,
        IFilmDetailsPageStateProps,
        IFilmDetailsPageDispatchProps {}

export const FilmDetailsPageContainer: FunctionComponent<
    IFilmDetailsPageContainerProps
> & { onInit: (store: any, match: any) => any } = ({
    history,
    match,
    fetchFilmDetailsInfoIfNeeded,
    isFilmDetailsLoading,
    isFilmsLoading,
    filmDetails = null,
    setParams,
    clearFilmDetails
}) => {
    useEffect(() => {
        const defaultParams = {
            search: '',
            searchBy: 'genres',
            sortBy: 'release_date',
            sortOrder: 'desc',
            limit: 50
        };

        setParams(defaultParams);

        fetchFilmDetailsInfoIfNeeded(match.params.id);
    }, [match.params.id, fetchFilmDetailsInfoIfNeeded, setParams]);

    useEffect(() => {
        if (!filmDetails) {
            // tslint:disable-next-line
            return () => {};
        }

        if (!Object.keys(filmDetails).length) {
            history.push('/not-found');
        }

        return () => clearFilmDetails();
    }, [filmDetails, history, clearFilmDetails]);

    return filmDetails ? (
        <FilmDetailsPage
            filmDetails={filmDetails}
            isFilmDetailsLoading={isFilmDetailsLoading}
            isFilmsLoading={isFilmsLoading}
        />
    ) : (
        <Loader />
    );
};

FilmDetailsPageContainer.onInit = (store, match) => {
    const defaultParams = {
        search: '',
        searchBy: 'genres',
        sortBy: 'release_date',
        sortOrder: 'desc',
        limit: 50
    };

    store.dispatch(paramsAction(defaultParams));

    store.dispatch(serverSideRenderedFlag(true));

    return store.dispatch(fetchFilmDetailsInfo(match.params.id));
};
