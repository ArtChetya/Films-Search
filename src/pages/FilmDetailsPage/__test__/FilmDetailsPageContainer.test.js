import { shallow } from 'enzyme/build';
import React from 'react';
import { FilmDetailsPageContainer } from '../FilmDetailsPageContainer';

jest.mock('../../../services/httpService');

describe('FilmDetailsPageContainer component', () => {
    let match;
    let history;
    let fetchFilmDetailsInfo;
    let isFilmDetailsLoading;
    let isFilmsLoading;
    let filmDetails;
    let setParams;
    let clearFilmDetails;

    beforeAll(() => {
        match = {
            params: { id: '1' }
        };

        history = {
            push: jest.fn
        };

        fetchFilmDetailsInfo = jest.fn();

        isFilmDetailsLoading = false;

        isFilmsLoading = false;

        filmDetails = {
            id: 181790,
            title: 'Journey 3: From the Earth to the Moon',
            tagline: '',
            vote_average: 0,
            vote_count: 7,
            release_date: '2018-12-31',
            poster_path:
                'https://image.tmdb.org/t/p/w500/98tbNloMBztAVnWpAznKKVUdi2O.jpg',
            overview:
                'Sean and Hank go on their biggest adventure yet, to the moon.',
            budget: 0,
            revenue: 0,
            genres: ['Action', 'Adventure', 'Fantasy', 'Science Fiction'],
            runtime: null
        };

        setParams = jest.fn();

        clearFilmDetails = jest.fn();
    });

    it('should be rendered correctly', () => {
        const component = shallow(
            <FilmDetailsPageContainer
                match={match}
                history={history}
                fetchFilmDetailsInfo={fetchFilmDetailsInfo}
                isFilmDetailsLoading={isFilmDetailsLoading}
                isFilmsLoading={isFilmsLoading}
                filmDetails={filmDetails}
                setParams={setParams}
                clearFilmDetails={clearFilmDetails}
            />
        );

        expect(component).toMatchSnapshot();
    });
});
