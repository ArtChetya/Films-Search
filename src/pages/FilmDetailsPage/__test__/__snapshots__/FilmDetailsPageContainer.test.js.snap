// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmDetailsPageContainer component should be rendered correctly 1`] = `
<FilmDetailsPage
  filmDetails={
    Object {
      "budget": 0,
      "genres": Array [
        "Action",
        "Adventure",
        "Fantasy",
        "Science Fiction",
      ],
      "id": 181790,
      "overview": "Sean and Hank go on their biggest adventure yet, to the moon.",
      "poster_path": "https://image.tmdb.org/t/p/w500/98tbNloMBztAVnWpAznKKVUdi2O.jpg",
      "release_date": "2018-12-31",
      "revenue": 0,
      "runtime": null,
      "tagline": "",
      "title": "Journey 3: From the Earth to the Moon",
      "vote_average": 0,
      "vote_count": 7,
    }
  }
  isFilmDetailsLoading={false}
  isFilmsLoading={false}
/>
`;
